### PR TITLE
[Feature] Support calculate confusion matrix and plot it.

### PR DESCRIPTION
**Pull request by mzr1996**
## Modification

Add the `ConfusionMatrix` metric and a command-line tool to calculate/plot the confusion matrix.

## Use cases (Optional)

**1. Basic Usage**
```python
>>> import torch
>>> from mmcls.evaluation import ConfusionMatrix
>>> y_pred = [0, 1, 1, 3]
>>> y_true = [0, 2, 1, 3]
>>> ConfusionMatrix.calculate(y_pred, y_true, num_classes=4)
tensor([[1, 0, 0, 0],
        [0, 1, 0, 0],
        [0, 1, 0, 0],
        [0, 0, 0, 1]])
>>> # plot the confusion matrix
>>> import matplotlib.pyplot as plt
>>> y_score = torch.rand((1000, 10))
>>> y_true = torch.randint(10, (1000, ))
>>> matrix = ConfusionMatrix.calculate(y_score, y_true)
>>> ConfusionMatrix().plot(matrix)
>>> plt.show()
```
**2. Use with Evalutor**
```python
>>> import torch
>>> from mmcls.evaluation import ConfusionMatrix
>>> from mmcls.structures import ClsDataSample
>>> from mmengine.evaluator import Evaluator
>>> data_samples = [
...     ClsDataSample().set_gt_label(i%5).set_pred_score(torch.rand(5))
...     for i in range(1000)
... ]
>>> evaluator = Evaluator(metrics=ConfusionMatrix())
>>> evaluator.process(data_samples)
>>> evaluator.evaluate(1000)
{'confusion_matrix/result': tensor([[37, 37, 48, 43, 35],
         [35, 51, 32, 46, 36],
         [45, 28, 39, 42, 46],
         [42, 40, 40, 35, 43],
         [40, 39, 41, 37, 43]])}
```
**3. Use command-line tool**
```shell
python tools/analysis_tools/confusion_matrix.py \
    configs/resnet/resnet50_8xb16_cifar10.py \
    https://download.openmmlab.com/mmclassification/v0/resnet/resnet50_b16x8_cifar10_20210528-f54bfad9.pth \
    --show
```
图片

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/evaluation/metrics/single_label.py [hunk 1]
```diff
@@ -584,3 +585,203 @@ class (N, C).
...
+
+        return matrix
+
+    @staticmethod
+    def plot(confusion_matrix: torch.Tensor,
+             include_values: bool = False,
+             cmap: str = 'viridis',
+             classes: Optional[List[str]] = None,
+             colorbar: bool = True,
+             show: bool = True):
```
Comment (Ezra-Yu) [inline mmcls/evaluation/metrics/single_label.py:723]: I would prefer to add a save path parameter here so that the result image can be saved on the command line when using linux

File: mmcls/evaluation/metrics/single_label.py [hunk 2]
```diff
@@ -584,3 +585,203 @@ class (N, C).
...
+        >>> # plot the confusion matrix
+        >>> import matplotlib.pyplot as plt
+        >>> y_score = torch.rand((1000, 10))
+        >>> y_true = torch.randint(10, (1000, ))
+        >>> matrix = ConfusionMatrix.calculate(y_score, y_true)
+        >>> ConfusionMatrix().plot(matrix)
+        >>> plt.show()
+        >>>
+        >>> # ------------------- Use with Evalutor -------------------
+        >>> from mmcls.structures import ClsDataSample
```
[the remark this answers is not in the document]
Comment (mzr1996) [reply mmcls/evaluation/metrics/single_label.py]: Done

File: mmcls/evaluation/metrics/single_label.py (hunk 1 above)
Comment (mzr1996) [reply mmcls/evaluation/metrics/single_label.py:716]: The method will output the figure, users can save the figure manually if they want.

**Comment by sibet-lb** [comment]
Hi, about the third method: use the command line tool. I can't find confusion_matrix.py in Analysis_tools.

**Comment by mzr1996** [comment]
> Hi, about the third method: use the command line tool. I can't find confusion_matrix.py in Analysis_tools. @mzr1996

I have moved it to the `analysis_tools` folder, you can find it now.

**Review by Ezra-Yu — APPROVED** [review]
LGTM.

**Comment by sibet-lb** [comment]
Hi, the github codebase is not updated. Another problem is, I used the latest version of mmcla, when I run confusion_matrix.py, an  error occured--cannot import name 'ConfusionMatrix' from 'mmcls.evaluation'. I found that 'ConfusionMatrix' is not defined, is there a script for 'ConfusionMatrix'? @mzr1996

**Comment by mzr1996** [comment]
> Hi, the github codebase is not updated. Another problem is, I used the latest version of mmcla, when I run confusion_matrix.py, an error occured--cannot import name 'ConfusionMatrix' from 'mmcls.evaluation'. I found that 'ConfusionMatrix' is not defined, is there a script for 'ConfusionMatrix'? @mzr1996

I think you can pull the latest `dev-1.x` branch and use it now.

**Comment by sibet-lb** [comment]
hi ， when I used the third way to get confusion matrix, the following error occured:
image
